### ISSUE TITLE
Is It Possible to patch the Softmax Function in FlexAttention?

### ISSUE THREAD

**Issue by yiliu30** [meta-pytorch/attention-gym#173]
Hi, 
I am experimenting with the Quant-Dequant on the prob tensor (the first input of the 2nd matmul in attention).
In the eager implementation, I tried modifying the code like this:
https://github.com/pytorch/pytorch/blob/4cc64d62346edb2e13320f20740f41ba41f64c70/torch/_higher_order_ops/flex_attention.py#L258
```python
def math_attention(
    query: torch.Tensor,
    key: torch.Tensor,
    value: torch.Tensor,
    score_mod: Callable,
    block_mask: tuple,
    scale: float,
    kernel_options: dict[str, Any],
    score_mod_other_buffers: tuple = (),
    mask_mod_other_buffers: tuple = (),
) -> tuple[torch.Tensor, torch.Tensor, torch.Tensor]:
    # broadcast query & key along head dim for GQA
    G = query.size(1) // key.size(1)
    value = torch.repeat_interleave(value, G, dim=1)
    key = torch.repeat_interleave(key, G, dim=1)

    Bq, Bkv = query.size(0), key.size(0)
    if not ((Bq == Bkv) or (Bq > 1 and Bkv == 1)):
        raise RuntimeError(f"Bq and Bkv must broadcast. Got Bq={Bq} and Bkv={Bkv}")

    key = key.expand((Bq, *key.size()[1:]))
    value = value.expand((Bq, *value.size()[1:]))

    _, post_mod_scores = _math_attention_inner(
        query,
        key,
        value,
        score_mod,
        block_mask,
        scale,
        kernel_options,
        score_mod_other_buffers,
        mask_mod_other_buffers,
    )

    # Set fully masked rows' sumexp to 0.0
    logsumexp = post_mod_scores.logsumexp(dim=-1)
    masked_rows = torch.all(post_mod_scores == -float("inf"), dim=-1)
    logsumexp = torch.where(masked_rows, -float("inf"), logsumexp)

    # working precision will be used so no need to cast to fp32
    max_scores = torch.max(post_mod_scores, dim=-1)[0]

    post_mod_scores = torch._safe_softmax(post_mod_scores, dim=-1)

    # NB: kernel computes in ln2 space, we always convert back at the top level op, so
    # for math impl we divide by log(2) because we will multiply by log(2)
    post_mod_scores = post_mod_scores.to(query.dtype)
    # post_mod_scores
    
    ## <<<<<<<<<<<<<<<< hook for quant-dequant testing >>>>>>>>>>>>>
    post_mod_scores = elemwise_quant_dequant(post_mod_scores)
    ## >>>>>>>>>>>>>>>>
    # post_mod_scores = post_mod_scores / post_mod_scores
    
    return (
        post_mod_scores @ value,
        logsumexp / math.log(2),
        max_scores / math.log(2),
    )

```

However, it seems like this change doesn't take effect after compiling `flex_attention`.
Is there an extension point or a part of the source code where I can apply this modification so it actually works? Thanks!

**Comment by drisspg** [comment]
Short answer is no, there is a nonzero amount of work needed to update the kernel implementation. It is possible but not without some larger changes
